server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshFailTests.java

```diff
@@ -0,0 +1,124 @@
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.RemoteTransportException;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class PostWriteRefreshFailTests extends IndexShardTestCase {
+
+    private final String errorMessage = "Unable to update to refresh an unpromotable shards";
+
+    private TransportService transportService;
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        transportService = MockTransportService.createNewService(Settings.EMPTY, Version.CURRENT, TransportVersion.CURRENT, threadPool);
+        transportService.start();
+        transportService.acceptIncomingRequests();
+        transportService.registerRequestHandler(
+            TransportUnpromotableShardRefreshAction.NAME,
+            ThreadPool.Names.SAME,
+            UnpromotableShardRefreshRequest::new,
+            (request, channel, task) -> channel.sendResponse(new IllegalStateException(errorMessage))
+        );
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        transportService.close();
+        super.tearDown();
+    }
+
+    public void testFailShardIfUnpromotableRequestFails() throws IOException {
+        allowShardFailures();
+
+        IndexShard shard = spy(newShard(true));
+        recoverShardFromStore(shard);
+        ReplicationGroup realReplicationGroup = shard.getReplicationGroup();
+        Engine.IndexResult result = indexDoc(shard, "_doc", "0");
+        ShardStateAction shardStateAction = mock(ShardStateAction.class);
+        PostWriteRefresh postWriteRefresh = new PostWriteRefresh(transportService, shardStateAction);
+
+        ReplicationGroup mockReplicationGroup = mock(ReplicationGroup.class);
+        when(shard.getReplicationGroup()).thenReturn(mockReplicationGroup).thenReturn(realReplicationGroup);
+
+        IndexShardRoutingTable routingTable = mock(IndexShardRoutingTable.class);
+        ShardRouting shardRouting = ShardRouting.newUnassigned(
+            shard.shardId(),
+            false,
+            RecoverySource.PeerRecoverySource.INSTANCE,
+            new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, "message"),
+            ShardRouting.Role.SEARCH_ONLY
+        );
+        when(routingTable.unpromotableShards()).thenReturn(List.of(shardRouting));
+        when(mockReplicationGroup.getRoutingTable()).thenReturn(routingTable);
+
+        doAnswer(invocation -> {
+            // Unfortunately, we have to verify the arguments manually in the invocation method since
```

Comment:
Did you consider using a plain `ShardStateAction` and a `CapturingTransport` instead of Mockito?